File: src/main/java/ru/octol1ttle/flightassistant/hud/impl/SpeedDisplay.java
package ru.octol1ttle.flightassistant.hud.impl;

import java.awt.Color;
import net.minecraft.client.font.TextRenderer;
import net.minecraft.client.gui.DrawContext;
import net.minecraft.text.Text;
import net.minecraft.util.math.MathHelper;
import ru.octol1ttle.flightassistant.Dimensions;
import ru.octol1ttle.flightassistant.DrawHelper;
import ru.octol1ttle.flightassistant.computers.impl.AirDataComputer;
import ru.octol1ttle.flightassistant.computers.impl.autoflight.AutoFlightController;
import ru.octol1ttle.flightassistant.config.FAConfig;
import ru.octol1ttle.flightassistant.hud.api.IHudDisplay;
import ru.octol1ttle.flightassistant.registries.ComputerRegistry;

public class SpeedDisplay implements IHudDisplay {
    private final Dimensions dim;
    private final AirDataComputer data = ComputerRegistry.resolve(AirDataComputer.class);
    private final AutoFlightController autoflight = ComputerRegistry.resolve(AutoFlightController.class);

    public SpeedDisplay(Dimensions dim) {
        this.dim = dim;
    }

    @Override
    public void render(DrawContext context, TextRenderer textRenderer) {
        int top = dim.tFrame;
        int bottom = dim.bFrame;

        int left = dim.lFrame - 2;
        int right = dim.lFrame;
        int unitPerPixel = 30;

        int floorOffset = MathHelper.floor(data.speed() * unitPerPixel);
        int yFloor = dim.yMid - floorOffset;

        int xSpeedText = left - 5;

        if (FAConfig.indicator().showSpeedReadout) {
            Color color = getSpeedColor(data.speed());
            DrawHelper.drawRightAlignedText(textRenderer, context, DrawHelper.asText("%.2f", data.speed()), xSpeedText, dim.yMid - 3, color);
            DrawHelper.drawBorder(context, xSpeedText - 29, dim.yMid - 5, 30, color);
        }

        if (FAConfig.indicator().showSpeedScale) {
            for (float i = 0; i <= 100; i += 0.25f) {
                int y = MathHelper.floor(dim.hScreen - i * unitPerPixel - yFloor);
                if (y < top || y > (bottom - 5))
                    continue;

                Color color = getSpeedColor(i);
                if (i % 1 == 0) {
                    DrawHelper.drawHorizontalLine(context, left - 2, right, y, color);
                    if (!FAConfig.indicator().showSpeedReadout || y > dim.yMid + 7 || y < dim.yMid - 7) {
                        DrawHelper.drawRightAlignedText(textRenderer, context, DrawHelper.asText("%.0f", i), xSpeedText, y - 3, color);
                    }
                }
                DrawHelper.drawHorizontalLine(context, left, right, y, color);
            }
        }
    }

    private Color getSpeedColor(float speed) {
        Integer targetSpeed = autoflight.getTargetSpeed();
        if (targetSpeed != null && Math.abs(targetSpeed - speed) <= 1.0f) {
            return FAConfig.indicator().advisoryColor;
        } else {
            return FAConfig.indicator().frameColor;
        }
    }

    @Override
    public void renderFaulted(DrawContext context, TextRenderer textRenderer) {
        DrawHelper.drawRightAlignedText(textRenderer, context, Text.translatable("short.flightassistant.speed"), dim.lFrame - 7, dim.yMid - 3, FAConfig.indicator().warningColor);
    }

    @Override
    public String getId() {
        return "speed";
    }
}

File: src/main/kotlin/ru/octol1ttle/flightassistant/impl/display/SpeedDisplay.kt
package ru.octol1ttle.flightassistant.impl.display

import kotlin.math.roundToInt
import net.minecraft.client.gui.DrawContext
import net.minecraft.text.Text
import net.minecraft.util.Identifier
import ru.octol1ttle.flightassistant.FlightAssistant
import ru.octol1ttle.flightassistant.api.computer.ComputerView
import ru.octol1ttle.flightassistant.api.display.Display
import ru.octol1ttle.flightassistant.api.display.HudFrame
import ru.octol1ttle.flightassistant.api.util.extensions.*
import ru.octol1ttle.flightassistant.config.FAConfig

class SpeedDisplay(computers: ComputerView) : Display(computers) {
    override fun allowedByConfig(): Boolean {
        return FAConfig.display.showSpeedReading || FAConfig.display.showSpeedScale
    }

    override fun render(drawContext: DrawContext) {
        with(drawContext) {
            val trueX: Int = HudFrame.left
            val trueY: Int = centerYI
            if (FAConfig.display.showSpeedReading) {
                renderSpeedReading(trueX, trueY)
            }
            if (FAConfig.display.showSpeedScale) {
                renderSpeedScale(trueX, trueY)
            }
        }
    }

    private fun DrawContext.renderSpeedReading(trueX: Int, trueY: Int) {
        matrices.push()
        val (x: Int, y: Int) = scaleMatrix(READING_MATRIX_SCALE, trueX, trueY)

        val speed: Double = computers.data.forwardVelocity.length() * 20
        val color: Int =
            if (speed <= 0.0) warningColor
            else primaryColor

        val text: String = speed.roundToInt().toString()
        val width: Int = getTextWidth(text) + 4
        val halfHeight = 6
        val textY: Int = y - 4

        drawBorder(x - width, y - halfHeight, width + 1, halfHeight * 2 - 1, color)
        drawRightAlignedText(text, x - 1, textY, color)

        matrices.pop()
    }

    private fun DrawContext.renderSpeedScale(x: Int, y: Int) {
        val speed: Double = computers.data.forwardVelocity.length() * 20
        val color: Int =
            if (speed <= 0.0) warningColor
            else primaryColor

        val minY: Int = HudFrame.top
        val maxY: Int = (y + fontHeight * (speed + 1)).toInt().coerceIn(minY - 1..HudFrame.bottom)

        drawVerticalLine(x, minY, maxY, color)

        enableScissor(0, minY, scaledWindowWidth, maxY + 1)

        enableScissor(0, minY, scaledWindowWidth, (if (FAConfig.display.showSpeedReading) y - 6 * READING_MATRIX_SCALE else maxY).toInt() + 1)
        drawHorizontalLine(x - 20, x, y, color)
        drawHorizontalLine(x - 35, x, minY, color)
        for (i: Int in speed.toInt()..speed.toInt() + 100) {
            if (!drawSpeedLine(x, y, i, speed, color)) {
                break
            }
        }
        disableScissor()

        enableScissor(
            0,
            (if (FAConfig.display.showSpeedReading) y + 5 * READING_MATRIX_SCALE else minY).toInt(),
            scaledWindowWidth,
            maxY + 1
        )
        drawHorizontalLine(x - 35, x, maxY, color)
        for (i: Int in speed.roundToInt() downTo 0) {
            if (!drawSpeedLine(x, y, i, speed, color)) {
                break
            }
        }
        disableScissor()

        disableScissor()
    }

    private fun DrawContext.drawSpeedLine(x: Int, y: Int, speed: Int, currentSpeed: Double, color: Int): Boolean {
        val textY: Int = (y + fontHeight * (currentSpeed - speed)).toInt()
        if (textY < HudFrame.top - 100 || textY > HudFrame.bottom + 100) {
            return false
        }
        drawHorizontalLine(x - 5, x, textY, color)
        if (speed % 5 == 0) {
            drawRightAlignedText(speed.toString(), x - 6, textY - 3, color)
        }

        return true
    }

    override fun renderFaulted(drawContext: DrawContext) {
        with(drawContext) {
            drawRightAlignedText(
                Text.translatable("short.flightassistant.speed"),
                HudFrame.left, centerYI - 5, warningColor
            )
        }
    }

    companion object {
        private const val READING_MATRIX_SCALE: Float = 1.5f
        val ID: Identifier = FlightAssistant.id("speed")
    }
}
